pennylane/io/qualtran_io.py

```diff
@@ -418,3 +586,337 @@ def compute_matrix(*params, **hyperparams):  # pylint: disable=no-self-argument
             raise MatrixUndefinedError
 
         return matrix
+
+
+def _split_qubits(registers, qubits):  # pylint: disable=redefined-outer-name
+    """Function from the Qualtran-Cirq interop module that splits the flat list of qubits into
+    a dictionary of appropriately shaped qubit arrays."""
+
+    qubit_regs = {}
+    base = 0
+    for reg in registers:
+        qubit_regs[reg.name] = np.array(qubits[base : base + reg.total_bits()]).reshape(
+            reg.shape + (reg.bitsize,)
+        )
+        base += reg.total_bits()
+    return qubit_regs
+
+
+def _ensure_in_reg_exists(  # pylint: disable=too-many-branches
+    bb: "qt.BloqBuilder",
+    in_reg: "_QReg",
+    qreg_to_qvar: Dict["_QReg", "qt.Soquet"],
+) -> None:
+    """Modified function from the Qualtran-Cirq interop module that takes care of qubit allocations,
+    split and joins to ensure `qreg_to_qvar[in_reg]` exists."""
+    from qualtran.cirq_interop._cirq_to_bloq import _QReg
+
+    all_mapped_qubits = {q for qreg in qreg_to_qvar for q in qreg.qubits}
+    qubits_to_allocate = [q for q in in_reg.qubits if q not in all_mapped_qubits]
+    if qubits_to_allocate:
+        n_alloc = len(qubits_to_allocate)
+        qreg_to_qvar[
+            _QReg(qubits_to_allocate, dtype=qt.QBit() if n_alloc == 1 else qt.QAny(n_alloc))
+        ] = bb.allocate(n_alloc)
+
+    if in_reg in qreg_to_qvar:
+        # This is the easy case when no split / joins are needed.
+        return
```

Comment:
Is there a case, where this would fail? Do we need to raise some `RuntimeError ` at that point?